Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,405 @@
+                    await Task.yield()
+                }
+            }
+
+            group.addTask {
+                try await Task.sleep(nanoseconds: UInt64(timeout.nanoseconds))
+                print("Throwing")
+                throw NIOAsyncEmbeddedEventLoopError.timeoutAwaitingFuture
+            }
+
+            do {
+                // This force-unwrap is safe: there are only two tasks, one will either return or hang, and the
+                // one that returns cannot be nil, only error. We then cancel the other Task, but we'll never
+                // ask for its result.
+                let result = try await group.next()!
+                group.cancelAll()
+                return result
+            } catch {
+                print("Threw")
+                group.cancelAll()
+                throw error
+            }
+        }
+    }
+
+    /// Runs the event loop and moves "time" forward to the given point in time, running any scheduled
+    /// tasks that need to be run.
+    ///
+    /// - Note: If `deadline` is before the current time, the current time will not be advanced.
+    public func advanceTime(to deadline: NIODeadline) async {
+        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
+            self.queue.async {
+                let newTime = max(deadline, self.now)
+
+                var tasks = CircularBuffer<EmbeddedScheduledTask>()
+                while let nextTask = self.scheduledTasks.peek() {
+                    guard nextTask.readyTime <= newTime else {
+                        break
+                    }
+
+                    // Now we want to grab all tasks that are ready to execute at the same
+                    // time as the first.
+                    while let candidateTask = self.scheduledTasks.peek(), candidateTask.readyTime == nextTask.readyTime {
+                        tasks.append(candidateTask)
+                        self.scheduledTasks.pop()
+                    }
+
+                    // Set the time correctly before we call into user code, then
+                    // call in for all tasks.
+                    self._now.store(nextTask.readyTime.uptimeNanoseconds)
+
+                    for task in tasks {
+                        task.task()
+                    }
+
+                    tasks.removeAll(keepingCapacity: true)
+                }
+
+                // Finally ensure we got the time right.
+                self._now.store(newTime.uptimeNanoseconds)
+
+                continuation.resume()
+            }
+        }
+    }
+
+    /// Executes the given function in the context of this event loop. This is useful when it's necessary to be confident that an operation
+    /// is "blocking" the event loop. As long as you are executing, nothing else can execute in this loop.
+    public func executeInContext<ReturnType: Sendable>(_ task: @escaping @Sendable () throws -> ReturnType) async throws -> ReturnType {
```

Comment:
I think we need more docu for this. IIUC this method is mainly to ensure that you can schedule multiple tasks at once, without side-effects getting into the order. However you execute before potentially scheduled work. 
In most cases, you don't want to actually run something here. However execute might indicate that.

By actually "blocking" the EL, you easily run into deadlocks here. Anyhow, it took me quite a moment to wrap my head around all of this. I guess more documentation will help here.

Reply:
Return type needs to be `Sendable` because we'll construct it in one thread and access it from another. Swift compiler will moan if we don't. :grin:

Reply:
Added some documentation